src/ddmd/dmangle.d

```diff
@@ -173,21 +180,123 @@ public:
     }
 
     ////////////////////////////////////////////////////////////////////////////
+    // writes a back reference number with base 10 using
+    //  characters starting from '0' for all digits but the
+    //  last digit. It uses characters starting from lastDigitBase.
+    // This allows detecting the end of the number as well as
+    //  having different type of back references determined by the
+    //  last digit.
+    final void writeBackRefPos(size_t pos, char lastDigitBase)
+    {
+        enum base = 10;
+        size_t mul = 1;
+        while (pos >= mul * base)
+            mul *= base;
+        while (mul >= base)
```

Comment:
> is this loop reinventing number-to-string?

Yes, but with a letter as the last digit to detect the end of the number and encode one of three types into it).